Make scopes directly callable

AltGr opened the pull request:
Addresses #277

Quite a few changes are included here, some of which have some extra
implications visible in the language:

- adds the `Scope of { -- input_v: value ... }` construct in the language

- handle it down the pipeline:
 * `ScopeCall` in the surface AST
 * `EScopeCall` in desugared and scopelang
 * expressions are now traversed to detect dependencies between scopes
 * transformed into a normal function call in dcalc

- defining a scope now implicitely defines a structure with the same name, with
  the output variables of the scope defined as fields. This allows us to type
  the return value from a scope call and access its fields easily.
  * the implications are mostly in surface/name_resolution.ml code-wise
  * the `Scope_out` struct that was defined in scope_to_dcalc is no longer
    needed/used and the fields are no longer renamed (changes some outputs; the
    explicit suffix for variables with multiple states is ignored as well)
  * one benefit is that disambiguation works just like for structures when there
    are conflicts on field names
  * however, it's now a conflict if a scope and a structure have the same name
    (side-note: issues with conflicting enum / struct names or scope variables /
    subscope names were silent and are now properly reported)

- you can consequently use scope names as types for variables as well. Writing
  literals is not allowed though, they can only be obtained by calling the
  scope.

## Still TODO

- [x] context variables are not handled properly at the moment

- [x] error handling on invalid calls

- [x] tests show a small error message regression

- [x] lots of examples will need tweaking to avoid scope/struct name or struct fields / output variable conflicts

- add a `->` syntax to make struct field access distinct from scope output var
  access, enforced with typing. This is expected to reduce confusion of users
  and add a little typing precision. (in a later PR)

- [x] document the new syntax & implications (tutorial, cheat-sheet)

- a consequence of the changes is that subscope variables also can now be typed.
  A possible future evolution / simplification would be to rewrite subscopes as
  explicit scope calls early in the pipeline. That could also allow to
  manipulate them as expressions (bind them in let-ins, return them...)

AltGr commented:
Notes for review:
- the first 3 patches are small improvements not directly related to the PR main purpose
- I am not sure about the tutorial update (48d788d74465259f7da2ade87c2aeabccc834fcf), it may be worth a small discussion to keep it clear and streamlined (also, local let-ins are missing there)

AltGr commented:
Please see https://github.com/AltGr/catala/commits/scopecall-al (more specifically https://github.com/AltGr/catala/commit/84755111ef1e62ebf9d1c5aa93f72fa74094e86d) for a demo.
There is a couple small fixes that I need to cleanup and add here.

AltGr commented:
Added the small fix and error handling improvement found while testing on the "aides au logement"

denismerigoux commented on compiler/surface/name_resolution.ml:
```diff
@@ -563,9 +669,18 @@ let get_def_key
+      | Some (SubScope (v, u)) -> v, u
+      | Some _ ->
+        Errors.raise_spanned_error pos
+          "Invalid access to input variable, '%s' is not a subscope"
```
Ok so in some error messages I had stylised the names and identifiers coming from the source code in yellow, could be worth doing the same here. I maintain we should hire an intern to port https://docs.rs/ariadne/0.1.3/ariadne/ to OCaml and integrate it to the Catala compiler.

denismerigoux commented on examples/aides_logement/prologue.catala_fr:
````diff
@@ -387,7 +387,7 @@ champ d'application ÉligibilitéAidePersonnaliséeLogement:
 
 ```catala-metadata
-déclaration énumération ÉligibilitéAllocationLogement:
+déclaration énumération TypeÉligibilitéAllocationLogement:
````
I suspect that now name collisions between scopes and enum/structs are sanctioned with an error message?

AltGr replied:
Yes! small rework of name maps used in Name Resolution also allowed better handling of variable/subscope and enum/struct name conflicts which are now all properly rejected.

denismerigoux commented on tests/test_bool/good/test_bool.catala_en:
```diff
@@ -15,7 +15,7 @@ scope TestBool:
 let TestBool :
   TestBool_in{"foo_in": unit → bool; "bar_in": unit → integer} →
-    TestBool_out{"foo_out": bool; "bar_out": integer} =
+    TestBool{"foo": bool; "bar": integer} =
```
missing a space `TestBool {...` in the printer?

denismerigoux commented on tests/test_bool/good/test_bool.catala_en:
````diff
@@ -39,6 +39,11 @@ $ catala Interpret -s TestBool
 
 ```catala-test-inline
 $ catala Scopelang 
+type TestBool = {
````
I would have put `struct` instead of `type` here ; I know this looks like Ocaml but we can afford the extra precision.

denismerigoux commented on tests/test_io/good/condition_only_input.catala_en:
```diff
@@ -19,10 +19,10 @@ $ catala Dcalc -s B
     let a.x : bool = error_empty ⟨true ⊢ false⟩ in
-    let result : A_out{"y_out": integer} = A (A_in {"x_in"= a.x}) in
-    let a.y : integer = result."y_out" in
+    let result : A{"y": integer} = A (A_in {"x_in"= a.x}) in
```
The dissymetry between `_in` as a suffix for the input and no suffix for the output is weird. But not sure what we can do about it.

denismerigoux reviewed (approved):
Thanks so much @AltGr for this massive piece of work that brings a long-awaited feature into the language. Sorry for the review wait :)

AltGr commented on compiler/scopelang/scope_to_dcalc.ml:
```diff
@@ -225,6 +237,53 @@ let rec translate_expr (ctx : 'm ctx) (e : 'm Ast.expr) :
+          let expr =
+            match str_field, expr with
+            | Some (_, (Ast.Reentrant, _)), None ->
+              Some (Expr.unbox (Expr.elit LEmptyError (mark_tany m pos)))
```
Hm looks like it indeed 🤔. But that would normally have been caught by the re-typing on dcalc. I'll add the justification anyway if I find this to be correct.

AltGr commented on compiler/scopelang/scope_to_dcalc.ml:
```diff
@@ -825,38 +868,38 @@ let translate_program (prgm : 'm Ast.program) : 'm Dcalc.Ast.program =
-      scope_ordering
-      (Bindlib.box Nil, decl_ctx)
+     ending with the top-level scope. The decl_ctx is filled in left-to-right
+     order, then the chained scopes aggregated from the right. *)
```
Yup you used not to need to pass the ctx from the left but this is now required to get the out structs correctly defined

AltGr commented on compiler/surface/parser.mly:
```diff
@@ -79,6 +79,12 @@ small_expression:
+| CARDINAL {
+  (Builtin Cardinal, Pos.from_lpos $sloc)
+}
+| LSQUARE l = separated_list(SEMICOLON, expression) RSQUARE {
```
That one I solved a while ago already, it's fine with just the last paren:
```catala
champ d'application CalculAidePersonnaliséeLogementLocatif
  sous condition date_courante >=@ |2022-07-01|:
  définition taux_composition_familiale égal à
    si nombre_personnes_à_charge = 0 alors
      selon situation_familiale_calcul_apl sous forme
      -- PersonneSeule: 2,83%
      -- Couple: 3,15%
    sinon si nombre_personnes_à_charge = 1 alors
      2,70%
    sinon si nombre_personnes_à_charge = 2 alors
      2,38%
    sinon si nombre_personnes_à_charge = 3 alors
      2,01%
    sinon si nombre_personnes_à_charge = 4 alors
      1,85%
    sinon si nombre_personnes_à_charge = 5 alors
      1,79%
    sinon si nombre_personnes_à_charge = 6 alors
      1,73%
    sinon
      1,73% -. 0,06% *. entier_vers_décimal de
        (nombre_personnes_à_charge - 6)
```

AltGr commented on compiler/scopelang/dune:
```diff
@@ -1,7 +1,9 @@
- (libraries utils dcalc ocamlgraph))
+ (libraries utils dcalc ocamlgraph)
+ (flags
+  (:standard -short-paths)))
```
https://github.com/CatalaLang/catala/pull/318#discussion_r951471940

AltGr commented on compiler/scopelang/scope_to_dcalc.ml:
```diff
@@ -31,6 +31,10 @@ type 'm scope_sig_ctx = {
+  scope_sig_in_fields :
+    (StructFieldName.t * Ast.io_input Marked.pos) ScopeVarMap.t;
+      (** Mapping between the input scope variables and the input struct fields.
+          The boolean is true for 'context' variables which need to be thunked. *)
```
Huh indeed, this is a leftover

AltGr commented:
Addressed most of your remarks. Summing up what remains:
- have a proper discussion about how to structure the hinge between name_resolution and typing 
- fragile ordering of struct fields
- discuss what to do with scopes with empty input or output (specific call syntax ? warnings ?)
- printing intermediate states: use `struct` in scopelang, `"_in"` vs `""` for in/out var naming
